xdsl/irdl.py

```diff
@@ -727,6 +721,196 @@ class OptSuccessorDef(SuccessorDef, OptionalDef):
 OptSuccessor: TypeAlias = Block | None
 VarSuccessor: TypeAlias = list[Block]
 
+_ClsT = TypeVar("_ClsT")
+_ParamT = TypeVar("_ParamT")
+
+
+class OpDefField(Generic[_ClsT]):
```

Comment:
I tried using those first, but they store constraints, whereas we want the original type, since the operands in isolation don't have the information to resolve everything. I think there was a problem with type variables, but I might be misremembering.